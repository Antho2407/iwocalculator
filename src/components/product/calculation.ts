import moment from "moment";

export interface Tuple {
  repaymentDate: string;
  principal: number;
  interest: number;
  total: number;
}

export type Calculation = (
  amount: number,
  duration: number,
  rate: number
) => Tuple[];

/* Decided to leave the two calculations as separated even if they very similar, because future other repayments could potentially be completely different.
With more use cases though, we could try to extract some of the common patterns into functions */
export const RCF: Calculation = (
  amount: number,
  duration: number,
  rate: number
) => {
  let currentDate = moment();

  return [...Array(duration)].map((_, index) => {
    const principal = amount / duration;
    const interest = (amount - index * principal) * rate;
    const tuple = {
      repaymentDate: moment().format("l"),
      principal,
      interest,
      total: principal + interest,
    };

    currentDate = moment(currentDate).add(1, "M");

    return tuple;
  });
};

export const BL: Calculation = (
  amount: number,
  duration: number,
  rate: number
) => {
  let currentDate = moment();

  return [...Array(duration)].map((_, index) => {
    const principal = amount / duration;
    const interest =
      (amount - index * principal) * rate + (index === 0 ? amount * 0.1 : 0);
    const tuple = {
      repaymentDate: moment().format("l"),
      principal,
      interest,
      total: principal + interest,
    };
    currentDate = moment(currentDate).add(1, "M");

    return tuple;
  });
};
