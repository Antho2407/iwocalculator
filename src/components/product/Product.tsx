import React, { useState } from "react";
import { ProductRoot, Label, Table } from "./Product.components";
import FieldRow from "../fieldRow/FieldRow";
import { Calculation, Tuple } from "./calculation";
import { Input } from "../input/Input.components";

export interface ProductProps {
  label: string;
  amount: number;
  duration: number;
  calculation: Calculation;
}

const Product: React.FC<ProductProps> = ({
  label,
  amount,
  duration,
  calculation,
}) => {
  const [rate, setRate] = useState(3);
  const handleRateChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
    setRate(parseInt(e.target.value));
  };
  const tuples = calculation(amount, duration, rate);
  const getTotals = (tuples: Tuple[]) => {
    let principal = 0;
    let interest = 0;
    let total = 0;

    tuples.forEach((tuple) => {
      principal += tuple.principal;
      interest += tuple.interest;
      total += tuple.total;
    });

    return { principal, interest, total };
  };
  const totals = getTotals(tuples);
  const format = (number: number) => number.toFixed(2);

  return (
    <ProductRoot>
      <Label>{label}</Label>
      <FieldRow label="Interest rate" info="(in %)">
        <Input
          type="number"
          onChange={handleRateChanged}
          value={rate}
          min={0}
          max={100}
        />
      </FieldRow>
      <Table>
        <thead>
          <tr>
            <th>Repayment date</th>
            <th>Principal</th>
            <th>Interest</th>
            <th>Total repayment</th>
          </tr>
        </thead>
        <tbody>
          {tuples.map((tuple, index) => (
            <tr key={`tuple_${index}`}>
              <td>{tuple.repaymentDate}</td>
              <td>{format(tuple.principal)}</td>
              <td>{format(tuple.interest)}</td>
              <td>{format(tuple.total)}</td>
            </tr>
          ))}
          <tr>
            <td>Total</td>
            <td>{format(totals.principal)}</td>
            <td>{format(totals.interest)}</td>
            <td>{format(totals.total)}</td>
          </tr>
        </tbody>
      </Table>
    </ProductRoot>
  );
};

export default Product;
