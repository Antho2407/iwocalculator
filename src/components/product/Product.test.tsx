import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import Product, { ProductProps } from "./Product";
import { BL } from "./calculation";

describe("<Product />", () => {
  let wrapper: ShallowWrapper<ProductProps>;

  describe("When the component is mounted", () => {
    it("should render successfully", () => {
      wrapper = shallow(
        <Product
          label="test-label"
          amount={10000}
          duration={4}
          calculation={BL}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
