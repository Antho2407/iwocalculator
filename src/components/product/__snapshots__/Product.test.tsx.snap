// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> When the component is mounted should render successfully 1`] = `
<styled.div>
  <styled.h3>
    test-label
  </styled.h3>
  <FieldRow
    info="(in %)"
    label="Interest rate"
  >
    <styled.input
      max={100}
      min={0}
      onChange={[Function]}
      type="number"
      value={3}
    />
  </FieldRow>
  <styled.table>
    <thead>
      <tr>
        <th>
          Repayment date
        </th>
        <th>
          Principal
        </th>
        <th>
          Interest
        </th>
        <th>
          Total repayment
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="tuple_0"
      >
        <td>
          04/02/2020
        </td>
        <td>
          2500.00
        </td>
        <td>
          1300.00
        </td>
        <td>
          3800.00
        </td>
      </tr>
      <tr
        key="tuple_1"
      >
        <td>
          05/02/2020
        </td>
        <td>
          2500.00
        </td>
        <td>
          225.00
        </td>
        <td>
          2725.00
        </td>
      </tr>
      <tr
        key="tuple_2"
      >
        <td>
          06/02/2020
        </td>
        <td>
          2500.00
        </td>
        <td>
          150.00
        </td>
        <td>
          2650.00
        </td>
      </tr>
      <tr
        key="tuple_3"
      >
        <td>
          07/02/2020
        </td>
        <td>
          2500.00
        </td>
        <td>
          75.00
        </td>
        <td>
          2575.00
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          10000.00
        </td>
        <td>
          1750.00
        </td>
        <td>
          11750.00
        </td>
      </tr>
    </tbody>
  </styled.table>
</styled.div>
`;
